bots: Access state in external Zulip server.

derAnfaenger opened the pull request:
This depends on [zulip/#7227](https://github.com/zulip/zulip/pull/7227)

derAnfaenger commented:
Pushed some more commits, so bots can now actually store state on a Zulip server.

derAnfaenger commented:
The flask botserver currently creates a new `ExternalBotHandler` instance on each call to handle message. This would lead to StateHandler being reinstantiated over and over, fetching a bot's full state every time. As a follow-up, we should make the flaskserver only keep one instance of each bot's handler for it's runtime.

showell commented on zulip_bots/zulip_bots/lib.py:
```diff
@@ -65,7 +66,8 @@ def put(self, key, value):
     def get(self, key):
         # type: (Text) -> Text
-        return self.demarshal(self.state_[key])
+        # Return a copy of the entry to prevent a bot accidentally modifying self.state_ on its own.
```
It seems like we could reorder the commits to have the JSON change happen first, so that we don't need to do the deepcopy here.

derAnfaenger replied:
The problem I see is when bots change the default encoding; The new encoding must be guaranteed to not return any reference to the old object. Maybe we can make an additional check for this?

derAnfaenger replied:
Hmm, but then again I can't see a case where that could occur, except for the original `lamba obj: obj`, which wouldn't work anymore anyway.

showell commented:
LGTM except for one comment that I made.

Can you open a separate issue for the flaskserver resinstantiation problem?

derAnfaenger commented:
@showell updated the PR (merged the two respective commits and removed the deepcopy). Created #147 for the flaskserver issue.

derAnfaenger commented:
Resolved the merged conflicts.

timabbott commented on zulip/zulip/__init__.py:
```diff
@@ -859,6 +859,38 @@ def create_user(self, request=None):
+            request=request,
+        )
+
+    def get_state(self, request=None):
```
Let's call these e.g. `update_bot_storage`.  We might also want to rename the endpoints to use the `user_storage` name, given what we're calling this feature in the client-side code...

timabbott commented on zulip_bots/zulip_bots/lib.py:
```diff
@@ -56,15 +56,17 @@ class StateHandler(object):
     def get(self, key):
         # type: (Text) -> Text
+        # Ensure demarshal() never returns a reference to state_ to prevent
+        # a bot accidentally modifying self.state_ on its own.
```
Is that a TODO, or something else?

derAnfaenger replied:
That's a leftover comment, will remove.

timabbott commented:
I posted a couple comments; the rename could potentially do in a follow-up commit. 

The other thing I'd like to do is only have the bot fetch the storage if the bot has some declaration on the class like `USES_STORAGE = True`, to avoid doing a round-trip for every bot just because some bots support storage.  That can be a follow-up issue though.

@showell feel free to merge once you're satisfied here.

derAnfaenger commented:
@timabbott @showell updated. 
- removed the superfluous comments, 
- renamed get_state() and update_state() to get_storage() and update_storage() (I found these just as meaningful as get_[bot/user]_state(), while shorter.
- added a commit to make the StateHandler optional. I found this to be a nice use case of our META object. Note that this *is* compatible with our embedded bots framework, where the StateHandler is still provided by default.

derAnfaenger commented:
As a follow-up, should we add the META/bot_details logic to EmbeddedBotHandler?

showell commented:
Merged!  Thanks @derAnfaenger!

derAnfaenger commented:
@showell is it possible you commented "merged" without merging?

timabbott commented:
I bet the push got blocked due to forgetting to rebase.  Happens to me sometimes too :).

showell commented:
I'm not sure what happened here, but I merged for real now.
